src/api/useFetch.js

```diff
@@ -17,6 +17,210 @@ import {
+ * @param {string} [cacheKey=JSON.stringify({fetchFunction, fetchParams})] - if not `null`, use to
+ * lookup and serve result from cache.
+ * @param {Function} [cacheCondition=null] - optional callback, takes resolved promise data as input
+ * and must be truthy to write entry to cache (implication: prevents unresolved promise from being
+ * cached)
+ * @returns {[any, {isLoading: boolean, error: string|null}, callable]} - array containing the
+ * fetched data, the state object, and a callback to re-fetch (data state is *not* re-initialised
+ * allowing consumers to differeniate between inital fetch and re-fetch).
+ */
+const _useFetch = ({
+  fetchFunction,
+  fetchParams = null,
+  fetchCondition = null,
+  errorMessage = 'error',
+  retryCondition,
+  retryIntervalSeconds = 10,
+  showSnackbar = true,
+  cacheCondition = null,
+  cacheKey = JSON.stringify({
+    fetchFunction: fetchFunction.name,
+    fetchParams: fetchParams === null ? null : normaliseObject(fetchParams),
+  }),
+}) => {
+  const [data, setData] = React.useState()
+  const [state, setState] = React.useState({
+    isLoading: true,
+    error: null,
+  })
+
+  // toggle state to refresh useFetch hook, callable is returned next to data and state
+  const [refresh, setRefresh] = React.useState(true)
+  const resetState = () => {
+    setState({
+      isLoading: true,
+      error: null,
+    })
+  }
+  React.useEffect(() => resetState(), [refresh])
+
+  const retryIntervalRef = React.useRef()
+  const isFetching = React.useRef(false)
+
+  const { enqueueSnackbar } = useSnackbar()
+
+  const clearIntervalFromRef = (ref) => {
+    clearInterval(ref.current)
+    ref.current = null
+  }
+
+  React.useEffect(() => {
+    if (fetchCondition && !fetchCondition()) return
+
+    // track whether hook is mounted
+    let mounted = true
+
+    const fetchData = async () => {
+      // prevent multiple fetch attempts if component is re-rendered while fetching
+      if (isFetching.current) return
+      isFetching.current = true
+
+      // prevent stale updates to avoid memory leaks
+      if (!mounted) return
+
+      if (cacheKey !== null) {
+        const cachedResult = cache.get(cacheKey)
+        if (cachedResult) {
+          if (cachedResult instanceof Promise) {
+            try {
+              const result = await cachedResult
+              setData(result)
+              setState({
+                isLoading: false,
+                error: null,
+              })
+
+            } catch (error) {
+              setState({
+                isLoading: false,
+                error: error.toString(),
+              })
+              showSnackbar && enqueueSnackbar(
+                errorMessage,
+                {
+                  ...errorSnackbarProps,
+                  details: error.toString(),
+                  onRetry: () => fetchData(),
+                }
+              )
+            }
+
+            isFetching.current = false
+            return
+
+          } else {
```

Comment:
nit: Drop `else` block and de-dent (you already returned in `then` block anyways)